### PR TITLE
Renaming of pipeline run secrets

### PR DESCRIPTION
**Pull request by rinckm**
### Description

Add support for renaming secrets when being copied to pipeline run namespaces. The desired name is provided as annotation of the original secret in the tenant namespace.

Steward clients can use this feature to provision secrets into Steward clusters (tenant namespaces) _for the lifetime of a single pipeline run only_, without the need to synchronize parallel pipeline run set-ups/clean-ups.

For each pipeline run a client creates the needed secrets with a _unique_ name that differs from the _original_ secret name and refers to those in the pipeline run spec. The original secret is stored in a secret store outside the Steward cluster. When a pipeline run is finished, the client immediately removes all the corresponding secrets (for security reasons). Not using the original secret names ensures that parallel pipeline runs requiring the same original secret do not interfere with each other.

Because the original secret name is used as the Jenkins credential ID which pipelines depend on, the Secret objects in pipeline run sandbox namespaces must have the original name. Clients provide the original secret name as annotation. Steward renames secrets based on the annotation when it copies them to pipeline run sandbox namespaces.

In case of secret name collisions in pipeline run sandbox namespaces the pipeline run will fail with result code `error_config`.

### Dependency release notes

N/A

### Submitter checklist

- [x] Change has been tested (on a back-end cluster)
  [Integration test added](https://github.com/SAP/stewardci-core/pull/165/commits/fa3626cb89d658e542162a9c80e9e4d604319231)

- [x] [changelog.yaml] with upgrade notes are prepared and appropriate for the audience affected by the change (users or developer, depending on the change).
- [x] Semantic version diffed against [last release][releases] and updated accordingly. In this project the version has to be maintained here:
  - [/charts/steward/Chart.yaml](https://github.com/SAP/stewardci-core/blob/master/charts/steward/Chart.yaml) (`version` and `appVersion`)

In case dependencies have been updated:
- [ ] n/a Links to external changelogs added since the last release of our component
- [ ] n/a Changelogs read thoroughly, potential impact described, upgrade notes prepared (if necessary)
- [ ] n/a Check if dependency update affects our semantic version increment.

### Reviewer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] The Pull Request title is understandable and reflects the changes well
- [x] The Pull Request description is understandable and well documented
- [x] 'Upgrade notes' are documented in changelog.yaml (if required)
- [x] [changelog.yaml] entry for this Pull Request has been added
  - [x] Changelog entry contains all required information

[changelog.yaml]: https://github.com/SAP/stewardci-core/changelog.yaml
[releases]: https://github.com/SAP/stewardci-core/releases

### CODE REVIEW

**Comment by alxsap** [comment]
In which scenario did we experience issues with (unrenamed) secrets?

**Comment by rinckm** [comment]
> In which scenario have we been experiencing issues with (unrenamed) secrets?

We don't have issues with this.
With this option you could create the same secret multiple times in the tenant with different names.
This is helpful if you only want to store the secret in the tenant for the duration of the pipeline run and remove it afterwards.
With the renaming annotation you can ensure the secret has the same name inside each pipeline run.
By creating the same secret in the tenant with different names for each pipeline run you avoid raise conditions if multiple pipeline runs run in parallel.

File: docs/secrets/Secrets.md [hunk 1]
```diff
@@ -172,6 +172,9 @@ The Kubernetes Credentials Provider Plugin requires Secret resource objects to b
 Details can be found on the [Examples page][jenkins_k8s_credential_provider_plugin_examples] of the plugin.
 
 When a pipeline gets executed in a transient sandbox namespace, the secrets listed in `spec.secrets` of the corresponding PipelineRun resource object are copied to the sandbox namespace with the same name.
+
+If a renaming is required you may add an annotation with the key `steward.sap.com/secret-rename-to` and the new name as the value. The secret will be available with the new name in the Jenkins. The new name must be a vaild kubernetes name. In the `spec.secrets` you need to use the original name of the secret as it is existing in the tenant namespace.
+
```
Comment (anbrsap) [inline docs/secrets/Secrets.md:177]:
```suggestion
It is also possible to rename the secret while it gets copied by providing the desired name as annotation `steward.sap.com/secret-rename-to` on the source secret. The desired name must be a valid Kubernetes Secret name and be unique within the sandbox namespace. In `spec.secrets` of pipeline runs the original secret name must be used to select secrets.
```

File: pkg/runctl/secret_manager.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+type secretManager struct {
+	secretHelper secrets.SecretHelper
+}
+
+// NewSecretManager creates secrets in the run namesapce
+func NewSecretManager(secretHelper secrets.SecretHelper) runifc.SecretManager {
+	return &secretManager{
+		secretHelper: secretHelper,
+	}
+}
```
Comment (anbrsap) [inline pkg/runctl/secret_manager.go:22]: See <https://github.com/golang/go/wiki/CodeReviewComments#interfaces>.

`NewSecretManager()` should return the concrete type `SecretManger` (make it public).

File: pkg/runctl/secret_manager.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+// CopyAll copies the required secrets based on a pipelineRun to the run namespace
```
Comment (anbrsap) [inline pkg/runctl/secret_manager.go:24]:
```suggestion
// CopyAll copies the required secrets of a pipeline run to the respective run namespace.
```

File: pkg/runctl/secret_manager.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	secretName := pipelineRun.GetSpec().JenkinsFile.RepoAuthSecret
+	if secretName == "" {
+		return "", nil
+	}
+	repoServerURL, err := pipelineRun.GetPipelineRepoServerURL()
+	if err != nil {
+		// TODO: this method should not modify the pipeline run -> must be handled elsewhere
+		pipelineRun.UpdateMessage(err.Error())
+		pipelineRun.UpdateResult(v1alpha1.ResultErrorContent)
+		return "", err
```
Comment (anbrsap) [inline pkg/runctl/secret_manager.go:65]: Ouch. This should be fixed! Actually only the reconciler should modify the PipelineRun resource object.

We may introduce a new error annotation like `pkg/errors/recoverability.go` that indicates the Steward error classification (`error_config`, `error_content`, `error_infra`).

File: pkg/runctl/secret_manager.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+func (s *secretManager) copySecrets(pipelineRun k8s.PipelineRun, secretNames []string, filter secrets.SecretFilter, transformers ...secrets.SecretTransformer) ([]string, error) {
+	storedSecretNames, err := s.secretHelper.CopySecrets(secretNames, filter, transformers...)
+	if err != nil {
+		klog.Errorf("Cannot copy secrets %s for [%s]. Error: %s", secretNames, pipelineRun.String(), err)
+		pipelineRun.UpdateMessage(err.Error())
+		if s.secretHelper.IsNotFound(err) || k8serrors.IsInvalid(err) || k8serrors.IsAlreadyExists(err) {
+			pipelineRun.UpdateResult(v1alpha1.ResultErrorContent)
+		} else {
+			pipelineRun.UpdateResult(v1alpha1.ResultErrorInfra)
+		}
```
Comment (anbrsap) [inline pkg/runctl/secret_manager.go:98]: Ditto. Don't modify the PipelineRun object here.

File: pkg/runctl/secret_manager_test.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		Secrets: []string{
+			"secret1",
+			"secret2",
+		},
+		ImagePullSecrets: []string{
+			"imagePullSecret1",
+			"imagePullSecret2",
+		},
+	}
+)
```
Comment (anbrsap) [inline pkg/runctl/secret_manager_test.go:34]: I don't like using global variables for test helpers. I would put them into a struct, and use a constructor function for initialization.

File: test/builder/secret_test.go [hunk 7]
```diff
@@ -24,3 +25,21 @@ func Test_SecretBasicAuth(t *testing.T) {
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "foo",
+			Namespace:   "bar",
+			Labels:      map[string]string{"jenkins.io/credentials-type": "usernamePassword"},
+			Annotations: map[string]string{api.AnnotationSecretRename: "newName1"},
+		},
+		Type: v1.SecretTypeOpaque,
+		StringData: map[string]string{"username": "baz",
+			"password": "secret1",
+		},
```
Comment (anbrsap) [inline test/builder/secret_test.go:41]:
```suggestion
		StringData: map[string]string{
			"username": "baz",
			"password": "secret1",
		},
```

File: changelog.yaml [hunk 8]
```diff
@@ -21,6 +21,12 @@
 - version: "NEXT"
   date: TBD
   changes:
+  - type: enhancement
+    impact: minor
+    title: Rename secrets for jfr if annotation is provided
```
Comment (alxsap) [inline changelog.yaml:26]: Description missing

Comment (alxsap) [reply changelog.yaml:26]:
```suggestion
    title: Allow renamed secrets in run namespaces if annotation is provided
```

File: pkg/k8s/secrets/transformers_test.go [hunk 9]
```diff
@@ -26,6 +26,59 @@ func Test_UniqueNameTransformer_WithNameSet(t *testing.T) {
 	assert.DeepEqual(t, expected, transformed)
 }
 
+func Test_RenameByAttributeTransformer(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (alxsap) [reply pkg/k8s/secrets/transformers_test.go:29]: See https://github.com/SAP/stewardci-core/pull/165/files#r537306168
```suggestion
func Test_RenameByAnnotationTransformer(t *testing.T) {
```

File: pkg/k8s/secrets/transformers.go [hunk 10]
```diff
@@ -20,6 +21,20 @@ func UniqueNameTransformer() SecretTransformer {
 	}
 }
 
+// RenameByAttributeTransformer returns a secret transformer function that sets
+// `metadata.name` to the name defined in `metadata.annotations` with the defined key
+// no rename is done if the annotation with the key does not exist
+// no rename is done if an empty key is provided
+// no rename is done if the value of the annotation with the defined key is the empty string
+func RenameByAttributeTransformer(key string) SecretTransformer {
```
[the remark this answers is not in the document]
Comment (rinckm) [reply pkg/k8s/secrets/transformers.go]: Done: https://github.com/SAP/stewardci-core/pull/165/commits/8c4ee6cc6082aaa4514b62094b79b6bd62d57bbf

File: changelog.yaml [hunk 11]
```diff
@@ -21,6 +21,12 @@
 - version: "NEXT"
   date: TBD
   changes:
+  - type: enhancement
+    impact: minor
+    title: Allow renamed secrets in run namespaces if annotation is provided
```
Comment (alxsap) [inline changelog.yaml:26]: description missing

File: pkg/k8s/secrets/transformers_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (rinckm) [reply pkg/k8s/secrets/transformers_test.go]: Done

File: pkg/runctl/secret_manager.go [hunk 12]
```diff
@@ -0,0 +1,102 @@
+package runctl
```
[the remark this answers is not in the document]
Comment (rinckm) [reply pkg/runctl/secret_manager.go]: See https://github.com/SAP/stewardci-core/pull/165/commits/87569ae719bc0d94f271f5212c92063b503d03a3

File: pkg/runctl/secret_manager.go (hunk 2 above)
Comment (rinckm) [reply pkg/runctl/secret_manager.go]: See https://github.com/SAP/stewardci-core/pull/165/commits/87569ae719bc0d94f271f5212c92063b503d03a3

**Comment by alxsap** [comment]
[This one](https://github.com/SAP/stewardci-core/pull/190) will fix the voter. Merge master to this PR, once #190 is merged.

File: pkg/runctl/secret_manager.go (hunk 4 above)
Comment (rinckm) [reply pkg/runctl/secret_manager.go]: See https://github.com/SAP/stewardci-core/pull/165/commits/38ac2f590d6f7fc13b6c3e4adec595e3b8660b85

File: pkg/runctl/secret_manager.go (hunk 5 above)
Comment (rinckm) [reply pkg/runctl/secret_manager.go]: See https://github.com/SAP/stewardci-core/pull/165/commits/38ac2f590d6f7fc13b6c3e4adec595e3b8660b85

File: pkg/runctl/secret_manager_test.go (hunk 6 above)
Comment (rinckm) [reply pkg/runctl/secret_manager_test.go]: See https://github.com/SAP/stewardci-core/pull/165/commits/4bda8a075cf98dcd166781f5b92c0fcb7e186d55

File: changelog.yaml (hunk 11 above)
Comment (rinckm) [reply changelog.yaml:56]: Done

File: pkg/runctl/run_manager_test.go [hunk 13]
```diff
@@ -1396,177 +1398,38 @@ func Test_RunManager_Start_DoesNotSetPipelineRunStatus(t *testing.T) {
 	mockPipelineRun.EXPECT().UpdateState(gomock.Any()).Times(0)
 }
 
-func Test_RunManager_Start_DoesCopySecret(t *testing.T) {
+func Test_copySecretsToRunNamespace_DoesCopySecret(t *testing.T) {
```
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1401]:
```suggestion
func Test_RunManager_copySecretsToRunNamespace_DoesCopySecret(t *testing.T) {
```

File: pkg/runctl/run_manager_test.go [hunk 14]
```diff
@@ -1396,177 +1398,38 @@ func Test_RunManager_Start_DoesNotSetPipelineRunStatus(t *testing.T) {
...
-	mockPipelineRun.EXPECT().UpdateResult(stewardv1alpha1.ResultErrorInfra)
-	mockPipelineRun.EXPECT().String() //logging
-
-	// EXERCISE
-	err := examinee.Start(mockPipelineRun, config)
-	assert.Assert(t, err != nil)
+	cloneSecret, imagePullSecrets, err := examinee.copySecretsToRunNamespace(runCtx)
+	assert.NilError(t, err)
+	assert.Equal(t, "cloneSecret1", cloneSecret)
+	assert.DeepEqual(t, []string{"foo", "bar"}, imagePullSecrets)
```
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1432]:
```suggestion
	cloneSecret, imagePullSecrets, resultError := examinee.copySecretsToRunNamespace(runCtx)

	// VERFIY
	assert.NilError(t, resultError)
	assert.Equal(t, "cloneSecret1", cloneSecret)
	assert.DeepEqual(t, []string{"foo", "bar"}, imagePullSecrets)
```

File: pkg/errors/error_class_test.go [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+		class api.Result
+	}{
+		{api.ResultErrorInfra},
+		{api.ResultErrorContent},
+		{api.ResultErrorConfig},
+	} {
+		t.Run(string(tc.class), func(t *testing.T) {
+
+			//EXERCISE
+			classifiedErr := Classify(err1, tc.class)
```
[the remark this answers is not in the document]
Comment (rinckm) [reply pkg/errors/error_class_test.go:35]: See d94a53c274519b7405f10e2b97c0965a1d5ef3b2

File: pkg/runctl/run_manager.go [hunk 16]
```diff
@@ -738,7 +661,6 @@ func (c *runManager) cleanup(ctx *runContext) error {
 
 	namespace := ctx.runNamespace
 	if namespace == "" {
-		//TODO: Don't store on resource as message. Add it as event.
```
[the remark this answers is not in the document]
Comment (rinckm) [reply pkg/runctl/run_manager.go]: This was by mistake. Added it again d94a53c274519b7405f10e2b97c0965a1d5ef3b2

**Review by anbrsap — APPROVED** [review]

**Pull request merged**